src/Microsoft.Health.Fhir.Shared.Api/Modules/FhirModule.cs

```diff
@@ -159,6 +159,10 @@ ResourceElement SetMetadata(Resource resource, string versionId, DateTimeOffset
             services.Add<ProvenanceHeaderBehavior>().Scoped().AsSelf().AsImplementedInterfaces();
             services.Add<ProvenanceHeaderState>().Scoped().AsSelf().AsImplementedInterfaces();
 
+            // Register pipeline behavior to check service permission for CUD actions on StructuredDefinition,ValueSet,CodeSystem, ConceptMap.
+
+            services.Add<ProfileResourcesBehaviour>().Scoped().AsSelf().AsImplementedInterfaces();
```

Comment:
These will be picked up by convention as transient. We don't have a "ReplaceAsImplementedInterfaces" function so it might be better to individually use:

```suggestion
            services.Add<ProfileResourcesBehaviour>()
            .Scoped()
            .AsSelf()
            .ReplaceService<IPipelineBehavior<CreateResourceRequest, UpsertResourceResponse>>()
            .ReplaceService<IPipelineBehavior<UpsertResourceRequest, UpsertResourceResponse>>()
            .ReplaceService<IPipelineBehavior<ConditionalCreateResourceRequest, UpsertResourceResponse>>()
            .ReplaceService<IPipelineBehavior<ConditionalUpsertResourceRequest, UpsertResourceResponse>>()
            .ReplaceService<IPipelineBehavior<DeleteResourceRequest, DeleteResourceResponse>>();
```

Reply:
Perhaps we should consider also removing this default?

https://github.com/microsoft/fhir-server/blob/d40d65ba376db41fb4dd0616b3428961aea8107b/src/Microsoft.Health.Fhir.Shared.Api/Modules/MediationModule.cs#L28-L32

Reply:
I removed that.
I also not sure why did I choose them to be scoped and not singleton. It's a common behavior for all requests, what's the point to recreate it all the time?


---
In reply to: [583244843](https://github.com/microsoft/fhir-server/pull/1666#discussion_r583244843) [](ancestors = 583244843)

Reply:
I'd say this purely depends on the expected lifetime within the PipelineBehavior